#949 Implement preferred starting direction for AI initial exploration

victorxup opened the pull request:
Improve AI starting exploration during first 10 turns (that number was hardcoded before, I didn't change this):
When on mission "explore ocean" starting ship now tends to aim itself to the direction, which is calculated basing on the surrounding of starting plot (how Europe tiles are situated, which Europe type they are).
The system is based on already implemented weight calculation system and should not break it.
The vessel still will choose next move basing not only on how close it is to new parameter "preferred direction", but also how many plots will be revealed and a bit of random.

**The result is:** 
On maps with big distance to Europe AI doesn't get lost in the ocean so often ;
-> reaches coast faster
-> first settlement is founded more quickly;
-> first settlement is in better location regarding distance to Europe

The implemented code can be improved a bit in part of performance, but it will increase complexity as well.
Also this code is executed only for the first several turns, so I think there is no need to bother.


**Remaining possible improvements (TODO):**
- teach AI to value tiles with unfavourable wind less than other tiles (it already succesfully avoids storm)
- maybe scale first 10 turns period depending on map size or game speed

Nightinggale commented:
Please work based on the develop branch rather than master. We have a protected master, which isn't used for development, but rather will stay as the newest stable release version.

johan-buret reviewed (dismissed):
We can't merge like this what amounts to a full rewrite of CvUnitAI.cpp and CvPlayer.cpp

victorxup commented:
> We can't merge like this what amounts to a full rewrite of CvUnitAI.cpp and CvPlayer.cpp

Honestly, I didn't get what you mean.
Could you please get more into details? I'm open to suggestions.

I consider this change to be quite minor, but useful.

Voker57 commented:
> > We can't merge like this what amounts to a full rewrite of CvUnitAI.cpp and CvPlayer.cpp
> 
> Honestly, I didn't get what you mean. Could you please get more into details? I'm open to suggestions.
> 
> I consider this change to be quite minor, but useful.

For some reason, your commit removes every line in these files and then adds them again. Not sure why this happened, maybe mass whitespace changes, but it should be fixed.

MrZorG33 commented:
>maybe mass whitespace changes, but it should be fixed.

Indeed, in xml files they began to use tabs instead of whitespace. could this have an effect?

victorxup commented:
> For some reason, your commit removes every line in these files and then adds them again. Not sure why this happened, maybe mass whitespace changes, but it should be fixed.

hmm, OK, I'll look into this.
It's weird that github diff doesn't show that...

Voker57 commented:
> > For some reason, your commit removes every line in these files and then adds them again. Not sure why this happened, maybe mass whitespace changes, but it should be fixed.
> 
> hmm, OK, I'll look into this. It's weird that github diff doesn't show that...

it does
![diff](https://github.com/We-the-People-civ4col-mod/Mod/assets/22776/637eda05-3d31-49b0-a441-126047b6b736)

johan-buret commented:
@MrZorG33 the changes were brought to XML files, which are not involved in any way in this pull request.

@victorxup Github flags every single line of CvUnitAI.cpp and CvPlayer.cpp as changed, for no reason at all. Reviewing the changes in local show that, indeed, the effective change are much more reasonable in scope, like you told us.

victorxup commented:
Any idea what could be the cause? Maybe git cherry-picking? 
Okay, I'll handle this, force-push and notify you

Nightinggale commented:
Is it possible that there is an issue with line endings as in LF vs CRLF? I did set line endings .gitattributes, but it doesn't look like it is working as intended.

victorxup committed:
#949 Implement preferred starting direction for initial exploration

Direction is calculated depending on surrounding of starting plot.
Now AI starting units will not lose so often during first turns and will try to explore in the best direction

victorxup commented:
> Github flags every single line of CvUnitAI.cpp and CvPlayer.cpp as changed, for no reason at all. Reviewing the changes in local show that, indeed, the effective change are much more reasonable in scope, like you told us.

Fixed. Force pushed correct diff.
Pull Request can be reviewed now

johan-buret commented:
It looks much better this way.

johan-buret reviewed (approved):
The first tests are pretty conclusive. Even with a large coastal distance, the AI European first ships do not get lost

victorxup commented:
> The first tests are pretty conclusive. Even with a large coastal distance, the AI European first ships do not get lost

I usually play with 12 tiles distance, and AI is ok with it.
The tricky part is the rare cases and how bug-proof is this change.
Maybe we can add more asserts? But which?

johan-buret commented:
@Nightinggale  or @devolution79  are more at ease with the movement system than I am.

victorxup committed:
Fix for case when civ has no starting plot (e.g. barbarian)

victorxup commented:
Just a kind reminder:
I suppose, this update can be merged to dev-branch and properly tested by community.
